homeassistant/components/kef/media_player.py

```diff
@@ -0,0 +1,270 @@
+    else:
+        hass.data[DOMAIN][unique_id] = media_player
+        async_add_entities([media_player], update_before_add=True)
+
+
+class KefMediaPlayer(MediaPlayerDevice):
+    """Kef Player Object."""
+
+    def __init__(
+        self,
+        name,
+        host,
+        port,
+        maximum_volume,
+        volume_step,
+        standby_time,
+        inverse_speaker_mode,
+        sources,
+        ioloop,
+    ):
+        """Initialize the media player."""
+        self._name = name
+        self._sources = sources
+        self._speaker = AsyncKefSpeaker(
+            host,
+            port,
+            volume_step,
+            maximum_volume,
+            standby_time,
+            inverse_speaker_mode,
+            ioloop=ioloop,
+        )
+
+        self._state = STATE_UNKNOWN
+        self._muted = None
+        self._source = None
+        self._volume = None
+        self._is_online = None
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    async def async_update(self):
+        """Update latest state."""
+        _LOGGER.debug("Running async_update")
+        try:
+            self._is_online = await self._speaker.is_online()
+            if self._is_online:
+                (
+                    self._volume,
+                    self._muted,
+                ) = await self._speaker.get_volume_and_is_muted()
+                state = await self._speaker.get_state()
+                self._source = state.source
+                self._state = STATE_ON if state.is_on else STATE_OFF
+            else:
+                self._muted = None
+                self._source = None
+                self._volume = None
+                self._state = STATE_OFF
+        except (ConnectionRefusedError, ConnectionError, TimeoutError) as err:
+            _LOGGER.debug("Error in `update`: %s", err)
+            self._state = STATE_UNKNOWN
+
+    @property
+    def volume_level(self):
+        """Volume level of the media player (0..1)."""
+        return self._volume
+
+    @property
+    def is_volume_muted(self):
+        """Boolean if volume is currently muted."""
+        return self._muted
+
+    @property
+    def supported_features(self):
+        """Flag media player features that are supported."""
+        return SUPPORT_KEF
+
+    @property
+    def source(self):
+        """Name of the current input source."""
+        return self._source
+
+    @property
+    def source_list(self):
+        """List of available input sources."""
+        return self._sources
+
+    @property
+    def available(self):
+        """Return if the speaker is reachable online."""
+        return self._is_online
+
+    @property
+    def unique_id(self):
+        """Return the device unique id."""
+        return f"{self._speaker.host}:{self._speaker.port}"
```

Comment:
IP adress isn't stable so isn't a good fit for unique_id.

It should also normally not be configurable.

https://developers.home-assistant.io/docs/en/entity_registry_index.html#unique-id-requirements

Reply:
Changed to `kef-{lat}-{long}` :)

Reply:
If more than one speaker is configured in the same home assistant instance, will the coordinates have enough resolution?

Reply:
Probably not. However, we don't have access to the serial or MAC address.

I see many integrations that use the `ip`, so that's why I did it.

Reply:
What integrations are that?

If we don't have a good unique_id, we shouldn't implement it.

Reply:
> What integrations are that?

I stand corrected, I can't find them anymore (perhaps my memory is fooling me).

What should we do then?

Just remove `unique_id`?

Reply:
Yeah, if we can't make a good id we need to leave it out.

It would be ok to use `getmac` to get a mac address for the ip address. If that doesn't work we can still fall back to return `None`.

Reply:
It seems to work fine!

See my changes in e7e64e8bfc22ae909b41269cc7f6879490d2fcf0.